### ISSUE TITLE
How to preserve whitespace characters when updating a key's value?

### ISSUE THREAD

**Issue by copyitright** [jqlang/jq#1971]
I am not 100% sure if this is a known issue, but i use jq-win32 (v1.6) and when i add/update a key's value and write the resulting output to a new file, the whitespace characters are removed/stripped from the file. I want to keep them intact. 

See https://jqplay.org/s/LY8so0wxbk for the problem and notice the differences.

Regards,
copyitright

**Comment by pkoppstein** [comment]
If you are referring to the spaces before and after the ":" in the key:value pairs, please note that these have no semantic significance in JSON and jq has no command-line option for inserting spaces around the ":".  However, I believe the following jq program will do what you want, assuming the input is formatted JSON as produced for example by jq.

    def jsonstring: "(((?<=\\\\)\")|[^\"])*";
    sub( "(?<k>^ *\"" + jsonstring + "\"): "; .k + " : " ) 

To be clear - you would use this in a separate invocation of jq, e.g. in your case, along the lines of:

    jq '.selectedUser+={"profile" : "aabc1f631b274fea85fabec6c3487ee1"}' input.json |
      jq -rR 'sub( "(?<k>^ *\"(((?<=\\\\)\")|[^\"])*\"): "; .k + " : " )'

**Comment by copyitright** [comment]
Thank you for the given insight and solutions/workarounds.

The reason why i was looking for a solution is because another command (cmd's FINDSTR) that i use in conjunction with JQ wasn't giving me the expected output.

After more intensive testing i came to the conclusion that FINDSTR had another method to look for specific strings:

**TL;DR**

(the previous/old method that i used which didn't work under all circumstances)
```
>NUL 2>&1 FINDSTR /IC:"\"displayName\" : \"!old_name!\"^" "!PROFILES!"
```
(improved method that finds all occurences of "displayName", regardless of whitespace character variations)
```
>NUL 2>&1 FINDSTR /IRC:"\"displayName\" *: *\"!old_name!\"^" "!PROFILES!"
```

The important and missing ingredient that resolved the problem was the /I**R**C command switch.


**Long story short:**

The only thing i am wonder now is if there will be a noticable performance difference when using FINDSTR RegEx over JQ RegEx, when encountering bigger json files.

My inner voice tells me JQ is far more optimized/efficient, but if the difference is only one or two seconds, you won't hear me complaing :+)

**Comment by copyitright** [comment]
Besides the above, i've encountered another problem (limitation). It doesn't seem that JQ ends with an EXIT / ERRORLEVEL code. These are very useful when working with cmd/batch scripts... I'm quite baffled.

**Comment by pkoppstein** [comment]
@copyitright - Here's a relevant excerpt from the jq manual:

> -e / --exit-status:

>  Sets the exit status of jq to 0 if the last output values was neither false nor null, 1 if the last output value was either false or null, or 4 if no valid result was ever produced. Normally jq exits with 2 if there was any usage problem or system error, 3 if there was a jq program compile error, or 0 if the jq program ran.

> Another way to set the exit status is with the halt_error builtin function.

**Comment by copyitright** [comment]
Yep, i must read between the lines more often... this changes a whole lot for the better. Thanks 👍 :-)
